### ISSUE TITLE
"--files-from" scans a lot of other files?

### ISSUE THREAD

**Issue by gregfr** [ncw/rclone#498]
Greetings

I'm trying to use the "--files-from" options to limit scanning disk access.

I provide a list of 10 files, but with the verbose option I see that rclone is scanning thousands of files, is it the normal behavior?

Thanks in advance

greg

rclone v1.25 - debian 8.2 kernel 2.6.32-39 i686
target is hubic

**Comment by ncw** [comment]
I'm in the process of fixing this...

If you try the latest beta you should find it works just fine.

http://pub.rclone.org/v1.29-33-g085677d%CE%B2/

**Comment by ncw** [comment]
I've double checked this is fixed so I'm going to close this now.

**Comment by gregfr** [comment]
Thanks for working on this.
I've download the version at the link you provided and still sees a long list of files. Maybe I'm missing something?

**Comment by ncw** [comment]
It will scan files in the directories that your files are in, but it shouldn't go into any other directories if that makes sense. That should make things much more efficient.

**Comment by gregfr** [comment]
I've also tried to compile the github current version and same thing...

`./rclone -v -n   --files-from=lst.txt sync /source/ hubic:target`

lst.txt contains only 2lines, but rclone outputs thousands of line for several minutes...

**Comment by gregfr** [comment]
it seems it's listing the whole /source/ disk

**Comment by ncw** [comment]
Here is what I tried in the rclone source directory

```
$ cat z
appveyor.yml
CONTRIBUTING.md
dircache
hubic/hubic.go
```

Then

```
clone -v --dry-run --files-from=z copy . /tmp/z
```

Which lists this

```
2016/06/04 15:10:23 rclone: Version "v1.29" starting with parameters ["rclone" "-v" "--dry-run" "--files-from=z" "copy" "." "/tmp/z"]
2016/06/04 15:10:23 Local file system at /tmp/z: Modify window is 1ns
2016/06/04 15:10:23 Local file system at /tmp/z: Not making directory as dry run is set
2016/06/04 15:10:23 Local file system at /home/ncw/Code/Go/src/github.com/ncw/rclone: Building file list
2016/06/04 15:10:23 cross-compile: Excluded from sync (and deletion)
2016/06/04 15:10:23 RELEASE.md: Excluded from sync (and deletion)
2016/06/04 15:10:23 make_manual.py: Excluded from sync (and deletion)
2016/06/04 15:10:23 Local file system at /tmp/z: Building file list
2016/06/04 15:10:23 redirect_stderr_unix.go: Excluded from sync (and deletion)
2016/06/04 15:10:23 rclone_test.go: Excluded from sync (and deletion)
2016/06/04 15:10:23 redirect_stderr.go: Excluded from sync (and deletion)
2016/06/04 15:10:23 rclone-listdir-v1.29-23-gdbfa703.zip: Excluded from sync (and deletion)
2016/06/04 15:10:23 z: Excluded from sync (and deletion)
2016/06/04 15:10:23 versioncheck.go: Excluded from sync (and deletion)
2016/06/04 15:10:23 README.md: Excluded from sync (and deletion)
2016/06/04 15:10:23 .travis.yml: Excluded from sync (and deletion)
2016/06/04 15:10:23 MANUAL.md: Excluded from sync (and deletion)
2016/06/04 15:10:23 Local file system at /tmp/z: Done building file list
2016/06/04 15:10:23 .gitignore: Excluded from sync (and deletion)
2016/06/04 15:10:23 Makefile: Excluded from sync (and deletion)
2016/06/04 15:10:23 rclone.1: Excluded from sync (and deletion)
2016/06/04 15:10:23 rclone.go: Excluded from sync (and deletion)
2016/06/04 15:10:23 MANUAL.html: Excluded from sync (and deletion)
2016/06/04 15:10:23 ISSUE_TEMPLATE.md: Excluded from sync (and deletion)
2016/06/04 15:10:23 COPYING: Excluded from sync (and deletion)
2016/06/04 15:10:23 hubic/hubic_test.go: Excluded from sync (and deletion)
2016/06/04 15:10:23 hubic/auth.go: Excluded from sync (and deletion)
2016/06/04 15:10:23 Local file system at /home/ncw/Code/Go/src/github.com/ncw/rclone: Done building file list
2016/06/04 15:10:23 Local file system at /tmp/z: Waiting for checks to finish
2016/06/04 15:10:23 appveyor.yml: Not copying as --dry-run
2016/06/04 15:10:23 CONTRIBUTING.md: Not copying as --dry-run
2016/06/04 15:10:23 hubic/hubic.go: Not copying as --dry-run
2016/06/04 15:10:23 Local file system at /tmp/z: Waiting for transfers to finish
```

This divides into 
```
     2016/06/04 15:10:23 appveyor.yml: Not copying as --dry-run
     2016/06/04 15:10:23 CONTRIBUTING.md: Not copying as --dry-run
     2016/06/04 15:10:23 hubic/hubic.go: Not copying as --dry-run
```

And

```
     2016/06/04 15:10:23 cross-compile: Excluded from sync (and deletion)
     2016/06/04 15:10:23 RELEASE.md: Excluded from sync (and deletion)
     2016/06/04 15:10:23 make_manual.py: Excluded from sync (and deletion)
     2016/06/04 15:10:23 redirect_stderr_unix.go: Excluded from sync (and deletion)
     2016/06/04 15:10:23 rclone_test.go: Excluded from sync (and deletion)
     2016/06/04 15:10:23 redirect_stderr.go: Excluded from sync (and deletion)
     2016/06/04 15:10:23 rclone-listdir-v1.29-23-gdbfa703.zip: Excluded from sync (and deletion)
     2016/06/04 15:10:23 z: Excluded from sync (and deletion)
     2016/06/04 15:10:23 versioncheck.go: Excluded from sync (and deletion)
     2016/06/04 15:10:23 README.md: Excluded from sync (and deletion)
     2016/06/04 15:10:23 .travis.yml: Excluded from sync (and deletion)
     2016/06/04 15:10:23 MANUAL.md: Excluded from sync (and deletion)
     2016/06/04 15:10:23 .gitignore: Excluded from sync (and deletion)
     2016/06/04 15:10:23 Makefile: Excluded from sync (and deletion)
     2016/06/04 15:10:23 rclone.1: Excluded from sync (and deletion)
     2016/06/04 15:10:23 rclone.go: Excluded from sync (and deletion)
     2016/06/04 15:10:23 MANUAL.html: Excluded from sync (and deletion)
     2016/06/04 15:10:23 ISSUE_TEMPLATE.md: Excluded from sync (and deletion)
     2016/06/04 15:10:23 COPYING: Excluded from sync (and deletion)
     2016/06/04 15:10:23 hubic/hubic_test.go: Excluded from sync (and deletion)
     2016/06/04 15:10:23 hubic/auth.go: Excluded from sync (and deletion)
```
So you can see it traversed the current directory and the `hubic` directory but none of the other directories.

Do you see a similar effect?

The fix should stop it scanning directories it doesn't need to but will still note the files in those directories.  

The latest beta is: http://pub.rclone.org/v1.29-48-ge2788aa%CE%B2/

**Comment by gregfr** [comment]
It's not what I'm seeing on my folder:

```
2016/06/07 20:22:45 bin/bash: Excluded from sync (and deletion)                                                       
2016/06/07 20:22:45 bin/bunzip2: Excluded from sync (and deletion)                                                    
2016/06/07 20:22:45 bin/bzcat: Excluded from sync (and deletion)                                                      
2016/06/07 20:22:45 bin/bzdiff: Excluded from sync (and deletion)
...
2016/06/07 20:22:45 bin/znew: Excluded from sync (and deletion)                                                       
2016/06/07 20:22:48 etc/.pwd.lock: Excluded from sync (and deletion)                                                  
2016/06/07 20:22:48 etc/adduser.conf: Excluded from sync (and deletion)                                               
2016/06/07 20:22:48 etc/aliases: Excluded from sync (and deletion)                                                    
2016/06/07 20:22:48 etc/alternatives/README: Excluded from sync (and deletion)                                        
2016/06/07 20:22:48 etc/apache2/apache2.conf: Excluded from sync (and deletion) 
```

I'll try with the source directory.

**Comment by gregfr** [comment]
It's working as you described on the local source folder; however, if the target is hubic, I see the same listing, which I guess is the remote listing...

**Comment by ncw** [comment]
Ah, that makes sense.  It is very inefficient with hubic (also swift, dropbox, b2, s3) to list individual directories so rclone does a complete list for those remotes.  This should end up being quicker overall, but does mean that you will get each file in a `-v` listing.

**Comment by gregfr** [comment]
Ok I understand but as I'm backing up a whole server, it's very long, even for a few files. Is there a way to speed things up?

**Comment by ncw** [comment]
The overhead is the listing of the remote container.

How long does `rclone size remote:` take?

And how long does the copy take with the `--files-from` option?

When I try that on my swift server it runs at about 10,000 files per second.

One way of speeding things up would be to split the files list up by directory, so if you originally started with

  * `rclone copy --files-list z / remote:`

with files list

```
/home/user/file1
/home/user2/dir2/file2
```

You could then do two rclones one for `/home/user` and another for `/home/user2/dir2`.

I've just made a ticket #517 which would be a better fix for this in the long run.

**Comment by ncw** [comment]
I'm going to close this now as the initial problem is fixed.

The proper fix is probably #517 which I encourage you to subscribe to.

Please feel free to re-open if you have more info to add.

Thanks

Nick

**Comment by gregfr** [comment]
Thanks I'll try to find time to do some compartive tests.
